src/fastapi_app/dependencies.py

```diff
@@ -0,0 +1,123 @@
+    OPENAI_CHAT_HOST = os.getenv("OPENAI_CHAT_HOST")
+    if OPENAI_EMBED_HOST == "azure":
+        openai_embed_deployment = os.getenv("AZURE_OPENAI_EMBED_DEPLOYMENT", "text-embedding-ada-002")
+        openai_embed_model = os.getenv("AZURE_OPENAI_EMBED_MODEL", "text-embedding-ada-002")
+        openai_embed_dimensions = int(os.getenv("AZURE_OPENAI_EMBED_DIMENSIONS", 1536))
+    else:
+        openai_embed_deployment = "text-embedding-ada-002"
+        openai_embed_model = os.getenv("OPENAICOM_EMBED_MODEL", "text-embedding-ada-002")
+        openai_embed_dimensions = int(os.getenv("OPENAICOM_EMBED_DIMENSIONS", 1536))
+    if OPENAI_CHAT_HOST == "azure":
+        openai_chat_deployment = os.getenv("AZURE_OPENAI_CHAT_DEPLOYMENT", "gpt-35-turbo")
+        openai_chat_model = os.getenv("AZURE_OPENAI_CHAT_MODEL", "gpt-35-turbo")
+    elif OPENAI_CHAT_HOST == "ollama":
+        openai_chat_deployment = "phi3:3.8b"
+        openai_chat_model = os.getenv("OLLAMA_CHAT_MODEL", "phi3:3.8b")
+    else:
+        openai_chat_deployment = "gpt-3.5-turbo"
+        openai_chat_model = os.getenv("OPENAICOM_CHAT_MODEL", "gpt-3.5-turbo")
+    return FastAPIAppContext(
+        openai_chat_model=openai_chat_model,
+        openai_embed_model=openai_embed_model,
+        openai_embed_dimensions=openai_embed_dimensions,
+        openai_chat_deployment=openai_chat_deployment,
+        openai_embed_deployment=openai_embed_deployment,
+    )
+
+
+async def get_azure_credentials() -> azure.identity.DefaultAzureCredential | azure.identity.ManagedIdentityCredential:
+    azure_credential: azure.identity.DefaultAzureCredential | azure.identity.ManagedIdentityCredential
+    try:
+        if client_id := os.getenv("APP_IDENTITY_ID"):
+            # Authenticate using a user-assigned managed identity on Azure
+            # See web.bicep for value of APP_IDENTITY_ID
+            logger.info(
+                "Using managed identity for client ID %s",
+                client_id,
+            )
+            azure_credential = azure.identity.ManagedIdentityCredential(client_id=client_id)
+        else:
+            azure_credential = azure.identity.DefaultAzureCredential()
+        return azure_credential
+    except Exception as e:
+        logger.warning("Failed to authenticate to Azure: %s", e)
+        raise e
+
+
+async def get_engine():
+    """Get the agent database engine"""
+    load_dotenv(override=True)
+    azure_credentials = await get_azure_credentials()
+    engine = await create_postgres_engine_from_env(azure_credentials)
+    return engine
+
+
+async def get_async_session(engine: Annotated[AsyncEngine, Depends(get_engine)]):
+    """Get the agent database"""
+    async_session_maker = async_sessionmaker(engine, expire_on_commit=False)
+    async with async_session_maker() as async_session:
+        yield async_session
+
+
+async def get_openai_chat_client():
+    """Get the OpenAI chat client"""
+    azure_credentials = await get_azure_credentials()
```

Comment:
It looks like this is now calling DefaultAzureCredential in multiple places. Generally we should try to create the credential once and then pass it around, as it can be slow to run. Is that possible?

Reply:
I intentionally did that authenticate each request separately but sure, I will look into making it called once.

Reply:
I did fix that one too here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/da5220c9431c9c8ec637157f86fc0f18d4fe989e